src/selectors/linked-nodes.js

```diff
@@ -17,36 +15,83 @@ export const getCentralNode = createSelector(
 );
 
 /**
- * Recursively search through the edges data for ancestor and descendant nodes
+ * Gets a map of nodeIDs to the visible source edges and target edges
  * @param {Array} edges
+ */
+export const getVisibleEdgesByNode = createSelector(
+  [getVisibleEdges],
+  edges => {
+    const sourceEdges = {};
+    const targetEdges = {};
+
+    for (const edge of edges) {
+      if (!sourceEdges[edge.target]) {
+        sourceEdges[edge.target] = [];
+      }
+
+      sourceEdges[edge.target].push(edge);
+
+      if (!targetEdges[edge.source]) {
+        targetEdges[edge.source] = [];
+      }
+
+      targetEdges[edge.source].push(edge);
+    }
+
+    return { sourceEdges, targetEdges };
+  }
+);
+
+/**
+ * Finds all visible successors for the given nodeID in given direction
+ * @param {string} nodeID
+ * @param {Object} edgesByNode
+ * @param {string} direction
+ * @param {?object} visited
+ * @param {?boolean} isStart
+ * @param {?Array} empty
+ */
+const findLinkedNodes = (
+  nodeId,
+  edgesByNode,
+  direction,
+  visited = {},
+  isStart = true,
+  empty = []
```

Comment:
What is `empty`? It gets set once on first run, then checked each time, but never modified. Is it even necessary?

Reply:
Ah yes fair point, originally I was returning and joining arrays but decided to optimise this with a single array to avoid GC thrashing since it's a lot of calls. I just now made another change since you're right that the cached empty array is no longer needed this way.